### PR TITLE
feat!: add `Definition::Primitive`

### PR DESCRIPTION
**Pull request by dj8yfo**
this pr takes a step back and attempts to add definitions to all types, supported by `BorshSchema`,
in order to allow remove comments like
```rust
    /// Zero-sized types should follow the convention of either providing a [Definition] or
    /// specifying `"nil"` as their [Declaration] for this method to work correctly.
```
from various methods

### CODE REVIEW

File: borsh/src/schema.rs [hunk 1]
```diff
@@ -391,6 +417,7 @@ where
         };
         add_definition(Self::declaration(), definition, definitions);
         T::add_definitions_recursively(definitions);
+        E::add_definitions_recursively(definitions);
```
Comment (dj8yfo) [inline borsh/src/schema.rs:420]: it probably was copypasted from `Option<T>`, so this line was skipped

File: borsh/src/schema/container_ext/max_size.rs [hunk 2]
```diff
@@ -326,6 +346,7 @@ mod tests {
         test_ok::<Option<()>>(1);
         test_ok::<Option<u8>>(2);
         test_ok::<core::result::Result<u8, usize>>(9);
+        test_ok::<core::result::Result<u8, Vec<u8>>>(1 + 4 + MAX_LEN);
```
Comment (dj8yfo) [inline borsh/src/schema/container_ext/max_size.rs:349]: current master produces this fail of test assertion:
```bash
---- schema::container_ext::max_size::tests::max_serialized_size_built_in_types stdout ----
thread 'schema::container_ext::max_size::tests::max_serialized_size_built_in_types' panicked at 'assertion failed: `(left == right)`
  left: `Ok(9)`,
 right: `Err(MissingDefinition("Vec<u8>"))`'

```

File: borsh/tests/test_schema_validate.rs [hunk 3]
```diff
@@ -48,7 +48,7 @@ fn validate_for_derived_types() {
     test_ok::<Empty>();
     test_ok::<Named>();
     test_ok::<Unnamed>();
-    test_ok::<BorshSchemaContainer>();
+    // test_ok::<BorshSchemaContainer>();
```
Comment (dj8yfo) [inline borsh/tests/test_schema_validate.rs:51]:
```bash
thread 'validate_for_derived_types' panicked at 'assertion failed: `(left == right)`
  left: `Ok(())`,
 right: `Err(MissingDefinition("string"))`'
 ```

Comment (dj8yfo) [reply borsh/tests/test_schema_validate.rs:51]: where `str` and `String` would be 
```rust
pub enum Definition {
    /// `special_encoding` of `true` indicates that not all 
    /// sequences of `elements` type, allowed by the type, are valid with respect
    /// to `Sequence` type being defined 
    Sequence { elements: Declaration, special_encoding: bool },
    ...
}
impl BorshSchema for String {
    #[inline]
    fn add_definitions_recursively(
        definitions: &mut BTreeMap<Declaration, Definition>,
    ) {
        let definition = Definition::Sequence {
            elements: u8::declaration(),
            special_encoding: true,
        };
        add_definition(Self::declaration(), definition, definitions);
        u8::add_definitions_recursively(definitions);
        
    }
    #[inline]
    fn declaration() -> Declaration {
        "string".into()
    }
}
```
or just
```
        let definition = Definition::Sequence {
            elements: u8::declaration(),
        };
```

File: borsh/src/schema/container_ext.rs [hunk 4]
```diff
@@ -1,8 +1,8 @@
 use super::{BorshSchemaContainer, Declaration, Definition, Fields};
 
-use max_size::is_zero_size;
-pub use max_size::SchemaMaxSerializedSizeError;
-pub use validate::SchemaContainerValidateError;
+pub use max_size::Error as SchemaMaxSerializedSizeError;
+use max_size::{is_zero_size, ZeroSizeError};
+pub use validate::Error as SchemaContainerValidateError;
```
Comment (dj8yfo) [inline borsh/src/schema/container_ext.rs:5]: it appears to be too verbose in its own module, where the context of which `Error` is meant is more narrow

**Review by frol — APPROVED** [review]

**Pull request merged**